src/librbd/api/Migrate.cc

```diff
@@ -0,0 +1,991 @@
+      "", false, op_work_queue, &on_create);
+  req->send();
+  int r = on_create.wait();
+  if (r < 0) {
+    lderr(m_cct) << "header creation failed: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  auto dst_image_ctx = new librbd::ImageCtx(m_dst_image_name, m_dst_image_id,
+                                            nullptr, m_dst_io_ctx, false);
+  dst_image_ctx->ignore_migrating = true; // XXXMG: hackish
+  r = dst_image_ctx->state->open(false);
+  if (r < 0) {
+    lderr(m_cct) << "failed to open newly created header: " << cpp_strerror(r)
+                 << dendl;
+    return r;
+  }
+
+  dst_image_ctx->owner_lock.get_read();
+  if (dst_image_ctx->exclusive_lock != nullptr &&
+      !dst_image_ctx->exclusive_lock->is_lock_owner()) {
+    C_SaferCond ctx;
+    dst_image_ctx->exclusive_lock->acquire_lock(&ctx);
+    dst_image_ctx->owner_lock.put_read();
+    r = ctx.wait();
+    if (r < 0) {
+      lderr(m_cct) << "error acquiring exclusive lock: " << cpp_strerror(r)
+                   << dendl;
+      return r;
+    }
+  } else {
+    dst_image_ctx->owner_lock.put_read();
+  }
+
+  SnapSeqs snap_seqs;
+
+  C_SaferCond on_snapshot_copy;
+  auto snapshot_copy_req = librbd::deep_copy::SnapshotCopyRequest<I>::create(
+      m_image_ctx, dst_image_ctx, m_image_ctx->op_work_queue, &snap_seqs,
+      &on_snapshot_copy);
+  snapshot_copy_req->send();
+  r = on_snapshot_copy.wait();
+  if (r < 0) {
+    lderr(m_cct) << "failed to copy snapshots: " << cpp_strerror(r) << dendl;
+    dst_image_ctx->state->close();
+    return r;
+  }
+
+  C_SaferCond on_set_head;
+  librbd::ParentSpec parent_spec = {};
+  uint64_t parent_overlap = 0;
+  auto set_head_req = librbd::deep_copy::SetHeadRequest<I>::create(
+      dst_image_ctx, size, parent_spec, parent_overlap, &on_set_head);
+  set_head_req->send();
+  r = on_set_head.wait();
+  if (r < 0) {
+    lderr(m_cct) << "failed to set image head: " << cpp_strerror(r) << dendl;
+    dst_image_ctx->state->close();
+    return r;
+  }
+
+  r = dst_image_ctx->state->close();
+  if (r < 0) {
+    lderr(m_cct) << "failed to close image: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  cls::rbd::MigrationSpec migration_spec = {cls::rbd::MIGRATION_TYPE_DST,
+                                            cls::rbd::MIGRATION_STATE_STARTED,
+                                            m_src_io_ctx.get_id(),
+                                            m_src_image_name, m_src_image_id,
+                                            snap_seqs, size, false};
+
+  r = cls_client::migration_set(&m_dst_io_ctx, util::header_name(m_dst_image_id),
+                                migration_spec);
+  if (r < 0) {
+    lderr(m_cct) << "failed to set migration header: " << cpp_strerror(r)
+                 << dendl;
+    return r;
+  }
+
+  return 0;
+}
+
+template <typename I>
+int Migrate<I>::copy_data() {
+  ldout(m_cct, 10) << dendl;
+
+  int r = m_image_ctx->operations->flatten(m_prog_ctx);
```

Comment:
Why not deep-copy it so that eventually the parent/child relationship could be restored?

Reply:
It might look confusing but this is actually not flatten. When a migration image is set as a parent, flatten returns after STATE_FLATTEN_OBJECTS, so it only triggers copyup for every object. And CopyupRequest is modified to distinguish "migrating" case and use deep_copy in this case.

So it is deep-copy. I thought this is a right way to do if we want the migration being run simultaneously with other clients writing to the image.

Reply:
The copy-up logic wouldn't copy all snapshot data from the "parent" image to this new migration image, it would only copy the HEAD revision. We definitely want to allow a client to continue to use the image while the migration is in-progress, so perhaps we just need to add a new "deep-copy"-like operation to the RPC(?). 

On a semi-related note, we also need to handle the IO path where a client attempts to read from a specific snapshot on the destination image and therefore the "read from parent" needs to get redirected to the matching snapshot in the source image. Which, in that case, means you can ignore my comment about removing the snap id mapping from the migration struct.

Reply:
@dillaman I guess there is some reason why "deep-copyup" should be implemented on the server and may not be on the client (as it is right now)? I suppose a potential corruption when exclusive lock feature is not used?

Reply:
Also I can see how it can be done taking src and dst are different objects (probably belonging different pools). Or may be I misunderstand what you mean by adding "deep-copy"-like operation to the RPC?

Reply:
... actually, that is a good argument for using a "really deep" copy-up since it would still work w/o the need for exclusive lock so long as the final object write was using an improved "deep_copyup" guard cls method (i.e. the new operation would need to detect that the current object's snap id start is less than the incoming write's snap context).

Reply:
What I meant by a new "deep-copy" RPC message was just a way to explicitly call out that we are trying to do a migration and not a "flatten" which has a different meaning. Any client that can support the new "migration" feature bit would support the new op, so there is no harm in doing it instead of overloading an existing op.  Perhaps "migrate" would be a better name.